addon/lib/webextension-channels.js

```diff
@@ -0,0 +1,147 @@
+
+const {getExtensionUUID} = Cu.import('resource://gre/modules/Extension.jsm', {});
+
+function createChannelForAddonId(name, addonId) {
+  const targetExtensionUUID = getExtensionUUID(addonId);
+  const baseURI = Services.io
+        .newURI(`moz-extension://${targetExtensionUUID}/_blank.html`, null, null);
+  const principal = Services.scriptSecurityManager
+        .createCodebasePrincipal(baseURI, { addonId });
+
+  const addonChromeWebNav = Services.appShell.createWindowlessBrowser(true);
+  const docShell = addonChromeWebNav.QueryInterface(Ci.nsIInterfaceRequestor) // eslint-disable-line new-cap
+        .getInterface(Ci.nsIDocShell);
+  docShell.createAboutBlankContentViewer(principal);
+  const window = docShell.contentViewer.DOMDocument.defaultView;
+  const addonBroadcastChannel = new window.BroadcastChannel(name);
+
+  return {
+    addonChromeWebNav,
+    addonBroadcastChannel
+  };
+}
+
+const WebExtensionChannel = Class({ // eslint-disable-line new-cap
+  implements: [Disposable],
+
+  initialize(targetAddonId) {
+    this.pingListeners = new Set();
+
+    this.targetAddonId = targetAddonId;
+
+    const {
+      addonChromeWebNav,
+      addonBroadcastChannel
+    } = createChannelForAddonId(TESTPILOT_TELEMETRY_CHANNEL, targetAddonId);
+
+    // NOTE: Keep a ref to prevent it from going away during garbage collection
+    // (or the BroadcastChannel will stop working).
+    this.addonChromeWebNav = addonChromeWebNav;
+    this.addonBroadcastChannel = addonBroadcastChannel;
+  },
+
+  dispose() {
+    this.addonBroadcastChannel.removeEventListener();
+    this.addonChromeWebNav.close();
+    this.pingListeners.clear();
+
+    this.addonBroadcastChannel = null;
+    this.addonChromeWebNav = null;
+  },
+
+  registerPingListener(callback) {
+    this.pingListeners.add(callback);
+
+    if (this.pingListeners.size >= 0) {
+      this.addonBroadcastChannel.addEventListener('message', this);
+    }
+  },
+
+  unregisterPingListener(callback) {
+    this.pingListeners.delete(callback);
+
+    if (this.pingListeners.size === 0) {
+      this.addonBroadcastChannel.removeEventListener('message', this);
+    }
+  },
+
+  handleEvent(event) {
+    if (event.data) {
+      this.notifyPing(event.data, {addonId: this.targetAddonId});
+    }
+  },
+
+  notifyPing(data, sender) {
+    for (let pingListener of this.pingListeners) { // eslint-disable-line prefer-const
+      try {
+        pingListener({
+          senderAddonId: sender.addonId,
+          testpilotPingData: data
+        });
+      } catch (err) {
+        console.error('Error executing pingListener', err); // eslint-disable-line no-console
+      }
+    }
+  }
+});
+
+
+let channels = {};
+
+module.exports = {
+  WebExtensionChannel,
+
+  // Update all the channels on init.
+  init() {
+    this.updateExperimentChannels();
+  },
+
+  // Drop refs to channels for garbage collection
+  destroy() {
+    channels = {};
```

Comment:
Continuing with the theme of this review, I'm not sure if this will free the channels for GC, how sure are you feeling about this / have you asked anybody on the DOM team about it?